### PR TITLE
fix aggregate function

### LINKED ISSUE #211: Broken aggregate function in main branch
### What happened + What you expected to happen

If I use hierarchicalforecast v0.3.0 there's a problem with "unbalanced" time series, see: https://github.com/Nixtla/hierarchicalforecast/issues/189

However, if you exclude this specific issue, it works fine.

If I use the latest version from the main branch (maybe after this commit? https://github.com/Nixtla/hierarchicalforecast/commit/c1072171e074470e97d343e3fbb0188f68509f5d) I have huge issues.

I have unbalanced time series.

If I set the new "is_balanced" parameter to False (the default behavior), the returned dataframe has many missing dates (ds). For some unique_ids I expect many years of data but I just get 2 rows.

If I set the new "is_balanced" parameter to True, I get a reshaping error.

The same code works with version 0.3.0. I had to revert to the released 0.3.0 version because the version in the main branch is not usable for me.

### Versions / Dependencies

hierarchicalforecast main/master branch
Python 3.10
Linux
pandas 2.0.2

### Reproduction script

`Y_df, S_df, tags = aggregate(Y_df, spec)`

Unfortunately, I can't share my data set.

### Issue Severity

None

### LINKED ISSUE #224: [utils]  aggregate doesn't work correctly with most recent version of the code
### What happened + What you expected to happen

Hi, 
Thanks for adding sparse versions of the reconcile methods and S matrix.
aggregate() function however doesn't seem to work correctly. It doesn't seem to do the balancing (numpy_balance), and neither creating the aggregated time series based on spec. 

pip version: (previous version)
image

main branch: 
image

### Versions / Dependencies

main branch 

### Reproduction script

 Code to reproduce the result : 
```
df = pd.DataFrame(data = {'cat1': ['a', 'a', 'a'], 'cat2': ['1', '2', '3'],
                         'y': [10, 20, 30], 'ds': ['2020-01-01', '2020-02-01', '2020-03-01']})
df.insert(0, 'country', 'COUNTRY')

spec = [['country'], ['country', 'cat1'], ['country','cat1', 'cat2']]

Y_df, S_df, tags = aggregate(df, spec)
Y_df

``` 

### Issue Severity

None

### PR DESCRIPTION
**Pull request by jmoralez**
* Changes the `aggregate` function to handle both balanced and unbalanced series in the same way
* Deprecates the `is_balanced` argument
* Adds a simple test case that fails in the current main branch

Fixes #211
Fixes #224

### CODE REVIEW

**Comment by jmoralez** [comment]
@candalfigomoro it's in [0.3.0](https://github.com/Nixtla/hierarchicalforecast/blob/v0.3.0/hierarchicalforecast/utils.py#L193).

**Comment by candalfigomoro** [comment]
@jmoralez Wasn't it introduced with commit c1072171e074470e97d343e3fbb0188f68509f5d after version 0.3.0?

**Comment by candalfigomoro** [comment]
> @candalfigomoro it's in [0.3.0](https://github.com/Nixtla/hierarchicalforecast/blob/v0.3.0/hierarchicalforecast/utils.py#L193).

@jmoralez You are right, I thought it was introduced with commit c1072171e074470e97d343e3fbb0188f68509f5d after version 0.3.0 but that commit only extended its usage.

File: hierarchicalforecast/utils.py [hunk 1]
```diff
@@ -229,71 +157,77 @@ def aggregate(
...
+    S[:, -1] = tags[bottom_key]
+    categories = list(tags.values())
+    try:
+        encoder = OneHotEncoder(categories=categories, sparse_output=sparse_s, dtype=np.float32)
+    except TypeError:  # sklearn < 1.2
+        encoder = OneHotEncoder(categories=categories, sparse=sparse_s, dtype=np.float32)    
+    S = encoder.fit_transform(S).T
     if sparse_s:
-        Agg = S_df.sparse.to_coo().tocsr()[:n_agg, :]
+        df_constructor = pd.DataFrame.sparse.from_spmatrix
```
Comment (FedericoGarza) [inline hierarchicalforecast/utils.py:224]: This looks to me very elegant @jmoralez. Nice!😍

**Review by FedericoGarza — APPROVED** [review]
Thanks for this @jmoralez! LGTM :)

**Pull request merged**